File: Wave.cs
// Wave.cs
using System;
using System.IO;
using System.Diagnostics;
using System.Runtime.InteropServices;

class WaveGenerator
{
    private float freq, amp, duration;
    private int sampleRate;
    private string waveType;
    private short[] samples;

    public WaveGenerator(float freq, float amp, float duration, int sampleRate, string waveType)
    {
        this.freq = freq;
        this.amp = amp;
        this.duration = duration;
        this.sampleRate = sampleRate;
        this.waveType = waveType;
        Generate();
    }

    private void Generate()
    {
        int numSamples = (int)(sampleRate * duration);
        samples = new short[numSamples];
        Random rand = new Random();
        for (int i = 0; i < numSamples; i++)
        {
            float t = (float)i / sampleRate;
            float phase = 2 * (float)Math.PI * freq * t;
            float val = 0;
            switch (waveType)
            {
                case "sine":
                    val = (float)Math.Sin(phase);
                    break;
                case "square":
                    val = (float)Math.Sin(phase) >= 0 ? 1 : -1;
                    break;
                case "sawtooth":
                    val = 2 * (phase / (2 * (float)Math.PI) - (float)Math.Floor(phase / (2 * (float)Math.PI) + 0.5f));
                    break;
                case "triangle":
                    val = 2 * (float)Math.Abs(2 * (phase / (2 * (float)Math.PI) - (float)Math.Floor(phase / (2 * (float)Math.PI) + 0.5f))) - 1;
                    break;
                case "noise":
                    val = (float)(rand.NextDouble() * 2 - 1);
                    break;
                default:
                    val = 0;
                    break;
            }
            samples[i] = (short)(amp * val * 32767);
        }
    }

    public void SaveWAV(string filename)
    {
        using var fs = new FileStream(filename, FileMode.Create);
        using var bw = new BinaryWriter(fs);
        int dataSize = samples.Length * 2;
        int byteRate = sampleRate * 2;
        // RIFF
        bw.Write(System.Text.Encoding.ASCII.GetBytes("RIFF"));
        bw.Write(36 + dataSize);
        bw.Write(System.Text.Encoding.ASCII.GetBytes("WAVE"));
        // fmt
        bw.Write(System.Text.Encoding.ASCII.GetBytes("fmt "));
        bw.Write(16);
        bw.Write((short)1);
        bw.Write((short)1);
        bw.Write(sampleRate);
        bw.Write(byteRate);
        bw.Write((short)2);
        bw.Write((short)16);
        // data
        bw.Write(System.Text.Encoding.ASCII.GetBytes("data"));
        bw.Write(dataSize);
        foreach (var s in samples)
        {
            bw.Write(s);
        }
    }

    public void Play(string filename)
    {
        try
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                Process.Start("cmd", $"/c start {filename}");
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                Process.Start("afplay", filename);
            else
            {
                string[] players = { "ffplay", "aplay", "mpg123" };
                bool started = false;
                foreach (var p in players)
                {
                    try
                    {
                        Process.Start(p, filename);
                        started = true;
                        break;
                    }
                    catch { }
                }
                if (!started) Console.WriteLine("No suitable audio player found.");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine($"Playback error: {e.Message}");
        }
    }

    static void Main(string[] args)
    {
        if (args.Length > 0)
        {
            // Simple CLI parsing (could use System.CommandLine for better)
            var dict = new System.Collections.Generic.Dictionary<string, string>();
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i].StartsWith("-") || args[i].StartsWith("--"))
                {
                    string key = args[i].TrimStart('-');
                    if (i + 1 < args.Length && !args[i + 1].StartsWith("-"))
                        dict[key] = args[++i];
                    else
                        dict[key] = "true";
                }
            }
            float freq = dict.ContainsKey("f") ? float.Parse(dict["f"]) : 440f;
            string wtype = dict.ContainsKey("t") ? dict["t"] : "sine";
            float amp = dict.ContainsKey("a") ? float.Parse(dict["a"]) : 0.8f;
            float dur = dict.ContainsKey("d") ? float.Parse(dict["d"]) : 2.0f;
            int rate = dict.ContainsKey("r") ? int.Parse(dict["r"]) : 44100;
            string output = dict.ContainsKey("o") ? dict["o"] : "output.wav";
            bool noPlay = dict.ContainsKey("no-play");
            var gen = new WaveGenerator(freq, amp, dur, rate, wtype);
            gen.SaveWAV(output);
            Console.WriteLine($"Saved to {output}");
            if (!noPlay) gen.Play(output);
        }
        else
        {
            Console.WriteLine("=== Waveform Audio Generator ===");
            Console.Write("Enter type (sine, square, sawtooth, triangle, noise, default: sine): ");
            string wtype = Console.ReadLine().Trim();
            if (string.IsNullOrEmpty(wtype)) wtype = "sine";
            Console.Write("Enter frequency (Hz, default: 440): ");
            float.TryParse(Console.ReadLine(), out float freq);
            if (freq <= 0) freq = 440;
            Console.Write("Enter amplitude (0.0-1.0, default: 0.8): ");
            float.TryParse(Console.ReadLine(), out float amp);
            if (amp < 0 || amp > 1) amp = 0.8f;
            Console.Write("Enter duration (seconds, default: 2.0): ");
            float.TryParse(Console.ReadLine(), out float dur);
            if (dur <= 0) dur = 2.0f;
            Console.Write("Enter sample rate (Hz, default: 44100): ");
            int.TryParse(Console.ReadLine(), out int rate);
            if (rate <= 0) rate = 44100;
            Console.Write("Enter output filename (default: output.wav): ");
            string fname = Console.ReadLine().Trim();
            if (string.IsNullOrEmpty(fname)) fname = "output.wav";
            if (!fname.EndsWith(".wav")) fname += ".wav";
            var gen = new WaveGenerator(freq, amp, dur, rate, wtype);
            gen.SaveWAV(fname);
            Console.WriteLine($"Saved to {fname}");
            Console.Write("Play? (y/n, default: y): ");
            string play = Console.ReadLine().Trim().ToLower();
            if (play != "n") gen.Play(fname);
        }
    }
}

File: Wave.java
// Wave.java
import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Random;

public class Wave {
    private float freq, amp, duration;
    private int sampleRate;
    private String waveType;
    private short[] samples;
    private Random rand = new Random();

    public Wave(float freq, float amp, float duration, int sampleRate, String waveType) {
        this.freq = freq;
        this.amp = amp;
        this.duration = duration;
        this.sampleRate = sampleRate;
        this.waveType = waveType;
        generate();
    }

    private void generate() {
        int numSamples = (int)(sampleRate * duration);
        samples = new short[numSamples];
        for (int i = 0; i < numSamples; i++) {
            float t = (float)i / sampleRate;
            float phase = 2 * (float)Math.PI * freq * t;
            float val = 0;
            switch (waveType) {
                case "sine":
                    val = (float)Math.sin(phase);
                    break;
                case "square":
                    val = Math.sin(phase) >= 0 ? 1 : -1;
                    break;
                case "sawtooth":
                    val = 2 * (phase / (2 * (float)Math.PI) - (float)Math.floor(phase / (2 * (float)Math.PI) + 0.5));
                    break;
                case "triangle":
                    val = 2 * (float)Math.abs(2 * (phase / (2 * (float)Math.PI) - (float)Math.floor(phase / (2 * (float)Math.PI) + 0.5))) - 1;
                    break;
                case "noise":
                    val = (float)(rand.nextDouble() * 2 - 1);
                    break;
                default:
                    val = 0;
            }
            samples[i] = (short)(amp * val * 32767);
        }
    }

    public void saveWAV(String filename) throws IOException {
        try (FileOutputStream fos = new FileOutputStream(filename);
             DataOutputStream dos = new DataOutputStream(fos)) {
            int dataSize = samples.length * 2;
            int byteRate = sampleRate * 2;
            // RIFF
            dos.writeBytes("RIFF");
            dos.writeInt(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(36 + dataSize).array());
            dos.writeBytes("WAVE");
            // fmt
            dos.writeBytes("fmt ");
            dos.writeInt(16);
            dos.writeShort(1);
            dos.writeShort(1);
            dos.writeInt(sampleRate);
            dos.writeInt(byteRate);
            dos.writeShort(2);
            dos.writeShort(16);
            // data
            dos.writeBytes("data");
            dos.writeInt(dataSize);
            for (short s : samples) {
                dos.writeShort(s);
            }
        }
    }

    public void play(String filename) {
        String os = System.getProperty("os.name").toLowerCase();
        try {
            if (os.contains("win")) {
                Runtime.getRuntime().exec(new String[]{"cmd", "/c", "start", filename});
            } else if (os.contains("mac")) {
                Runtime.getRuntime().exec(new String[]{"afplay", filename});
            } else {
                String[] players = {"ffplay", "aplay", "mpg123"};
                for (String p : players) {
                    try {
                        Runtime.getRuntime().exec(new String[]{p, filename});
                        return;
                    } catch (IOException e) {}
                }
                System.out.println("No suitable audio player found.");
            }
        } catch (Exception e) {
            System.out.println("Playback error: " + e.getMessage());
        }
    }

    public static void main(String[] args) throws IOException {
        if (args.length > 0) {
            // Simple CLI
            java.util.Map<String, String> params = new java.util.HashMap<>();
            for (int i = 0; i < args.length; i++) {
                if (args[i].startsWith("-") || args[i].startsWith("--")) {
                    String key = args[i].replaceFirst("^-+", "");
                    if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
                        params.put(key, args[++i]);
                    } else {
                        params.put(key, "true");
                    }
                }
            }
            float freq = params.containsKey("f") ? Float.parseFloat(params.get("f")) : 440f;
            String wtype = params.getOrDefault("t", "sine");
            float amp = params.containsKey("a") ? Float.parseFloat(params.get("a")) : 0.8f;
            float dur = params.containsKey("d") ? Float.parseFloat(params.get("d")) : 2.0f;
            int rate = params.containsKey("r") ? Integer.parseInt(params.get("r")) : 44100;
            String output = params.getOrDefault("o", "output.wav");
            boolean noPlay = params.containsKey("no-play");
            Wave gen = new Wave(freq, amp, dur, rate, wtype);
            gen.saveWAV(output);
            System.out.println("Saved to " + output);
            if (!noPlay) gen.play(output);
        } else {
            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
            System.out.println("=== Waveform Audio Generator ===");
            System.out.print("Enter type (sine, square, sawtooth, triangle, noise, default: sine): ");
            String wtype = reader.readLine().trim();
            if (wtype.isEmpty()) wtype = "sine";
            System.out.print("Enter frequency (Hz, default: 440): ");
            float freq = 440f;
            try { freq = Float.parseFloat(reader.readLine()); } catch (NumberFormatException e) {}
            System.out.print("Enter amplitude (0.0-1.0, default: 0.8): ");
            float amp = 0.8f;
            try { amp = Float.parseFloat(reader.readLine()); } catch (NumberFormatException e) {}
            if (amp < 0 || amp > 1) amp = 0.8f;
            System.out.print("Enter duration (seconds, default: 2.0): ");
            float dur = 2.0f;
            try { dur = Float.parseFloat(reader.readLine()); } catch (NumberFormatException e) {}
            if (dur <= 0) dur = 2.0f;
            System.out.print("Enter sample rate (Hz, default: 44100): ");
            int rate = 44100;
            try { rate = Integer.parseInt(reader.readLine()); } catch (NumberFormatException e) {}
            System.out.print("Enter output filename (default: output.wav): ");
            String fname = reader.readLine().trim();
            if (fname.isEmpty()) fname = "output.wav";
            if (!fname.endsWith(".wav")) fname += ".wav";
            Wave gen = new Wave(freq, amp, dur, rate, wtype);
            gen.saveWAV(fname);
            System.out.println("Saved to " + fname);
            System.out.print("Play? (y/n, default: y): ");
            String play = reader.readLine().trim().toLowerCase();
            if (!play.equals("n")) gen.play(fname);
        }
    }
}
